node/src/components/linear_chain_sync/operations.rs

```diff
@@ -279,6 +307,116 @@ where
+                warn!(
+                    height,
+                    tag = ?BlockHeaderWithMetadata::TAG,
+                    ?peer,
+                    "Peer timed out",
+                );
+                // Peer timed out fetching the item, continue with the next peer
+                continue;
+            }
+            Err(error) => return Err(error),
+        }
+    }
+    Ok(None)
+}
+
+/// Fetches a block from the network by height.
+async fn fetch_and_store_block_by_height<REv, I>(
+    effect_builder: EffectBuilder<REv>,
+    height: u64,
+    trusted_validator_weights: &BTreeMap<PublicKey, U512>,
+    finality_threshold_fraction: Ratio<u64>,
+) -> Result<Option<Box<BlockWithMetadata>>, FetcherError<BlockWithMetadata, I>>
+where
+    REv: From<FetcherRequest<I, BlockWithMetadata>>
+        + From<NetworkInfoRequest<I>>
+        + From<StorageRequest>,
+    I: Eq + Debug + Clone + Send + 'static,
+{
+    for peer in effect_builder.get_peers_in_random_order().await {
+        match effect_builder
+            .fetch::<BlockWithMetadata, I>(height, peer.clone())
+            .await
+        {
+            Ok(FetchedData::FromStorage { item }) => return Ok(Some(item)),
+            Ok(FetchedData::FromPeer { item, .. }) => {
+                let BlockWithMetadata {
+                    block,
+                    finality_signatures,
+                } = *item.clone();
+
+                if *block.hash() != block.header().hash() {
+                    warn!(
+                        ?block,
+                        actual_block_header_hash = ?block.header().hash(),
+                        ?peer,
+                        "Block from peer did not have correct block header hash.",
+                    );
+                    // TODO: ban peer
+                    continue;
+                }
+
+                if block.body().hash() != *block.header().body_hash() {
+                    warn!(
+                        ?block,
+                        actual_block_body_hash = ?block.body().hash(),
+                        ?peer,
+                        "Block from peer did not have correct block body hash.",
+                    );
+                    // TODO: ban peer
+                    continue;
+                }
+
+                if let Err(error) = validate_finality_signatures(
+                    block.header(),
+                    &trusted_validator_weights,
+                    finality_threshold_fraction,
+                    &finality_signatures,
+                ) {
+                    warn!(
+                        ?error,
+                        ?peer,
+                        "Error validating finality signatures from peer.",
+                    );
+                    // TODO: ban peer
+                    continue;
+                }
+
+                // Store the block
+                // effect_builder
+                //     .put_block_to_storage(Box::new(block.clone()))
+                //     .await;
```

Comment:
Why not?

Reply:
I am calling legacy sync after this loop.  Sadly, legacy sync makes some assumptions if it finds blocks in the database.  So this will have to wait until fast sync can fly on its own.